### PR TITLE
Advanced Search

### LINKED ISSUE #30: Advanced search
## Details
The website could really use an advanced search feature. Currently you can only search for (parts of) titles.
Advanced search should include the ability to search for difficulty, number of downloads, rating, up-/down-votes and map type (Standard, Lightshow, ...). 

## Preferred Solution
On https://beatsaver.com/search in addition to the simple text input field for the title search i'd like to see a few more things added. These could be on a secondary search site or just be hidden behind some CSS/JS to hide them by default.

1. A bunch of checkboxes, one for each difficulty. By default all these checkboxes should be checked so a search just by title would always find all difficulties. The search should only return songs matching the difficulties of the checked checkboxes.

2. For map rating either two text input fields (lower and upper limit) or a range slider of some sorts (with a selector for the lower and a selector for the upper limit) would be good. Lower limit would be 0% and upper limit 100%, I guess. By default the search range would be between 0% and 100%.
 
3. For number of downloads and number of up-/down-votes a similar approach using the two text input boxes could be used. A slider is kind of not an option here though, since the upper limit is variable.
 
4. And lastly a number of checkboxes for the map types (Standard, Lightshow, and whatever other types there are) would be very nice to have.

### PR DESCRIPTION
**Pull request by adrianlungu**
## Proposed Changes
Fixes #30 

## Platforms
This pull request modifies: *(select all that apply)*
- [X] Client
- [X] Server

## Types of Changes
This pull request is contains: *(select all that apply)*
- [ ] Bug fixes *(non-breaking change which fixes an issue)*
- [X] New features *(non-breaking change which adds functionality)*
- [X] Breaking changes *(fix or feature that would cause existing functionality to not work as expected)*

## Checklist
- [X] I have read the [contribution guidelines](https://github.com/lolPants/beatsaver-reloaded/blob/master/.github/CONTRIBUTING.md)
- [?] I have checked the changes adhere to the project's style guide and all tests pass
- [X] I have (to the best of my ability) checked for vulnerabilities, or any ways that my code could be abused and concluded that it is safe to run in production

## Further Comments

**ATTENTION**
To be able to use ES sorting in the search, the Stats had to be indexed in ES as well. Therefor, the index required deletion and reindexing of the Beatmap data from Mongo. 

I've added a check to see if the index exists, otherwise it tries to recreate it. The deletion still has to be done manually before the server is started.

### CODE REVIEW

**Comment by lolPants** [comment]
Please resolve merge conflicts with a rebase or by submitting another PR with cherry-picked commits.

**Comment by lolPants** [comment]
Please resolve build failures.

**Comment by LeoDJ** [comment]
Is this pull request still active?

I'm asking because I was just looking into implementing an advanced search too, when stumbling over this PR.

If it's just some minor things that need to be fixed here, maybe I can do that too? Because I _really_ want an advanced search.

**Comment by adrianlungu** [comment]
I've been planning on wrapping this up but didn't get the chance, I'll try too wrap it up now; I'll let you know if I don't get the chance to finish in case you wanna continue working on it.

**Comment by adrianlungu** [comment]
Please test this a bit, while it works fine on my end, there seem to have been a few changes made to beatmaps and all.

Also, I made the following change in the server search.ts file:
```
//must: { query_string: { query } },
must: {query_string: {query: q}},
```

For some reason search wasn't working well for me using all the fields added to the query. If I get a chance to look into it more I'll come back with more details, if not, please feel free to enable the previous way it was working if without fields it behaves weirdly (it seemed fine on my end but I don't have a very large beatmap collection locally).

**Comment by lolPants** [comment]
The autolinter isn't working right now but I can tell that you've not linted the project.
Run `yarn run test` in both `client` and `server` directories to show lint failures.

I would also recommend installing the [TSLint Extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-tslint-plugin) for VSCode (or find one for whatever editor you are using) to show these inline in the IDE.

File: client/package.json [hunk 1]
```diff
@@ -80,21 +86,24 @@
...
     "dotenv": "^8.0.0",
-    "file-saver": "^2.0.2",
     "he": "^1.2.0",
     "koa": "^2.7.0",
     "koa-router": "^7.4.0",
     "koa-send": "^5.0.0",
     "koa-static": "^5.0.0",
-    "mongoose": "^5.6.9",
+    "mongoose": "^5.7.7",
+    "react-select": "^3.0.4",
```
Comment (lolPants) [inline client/package.json]: This should be a devDependency

File: client/package.json [hunk 2]
```diff
@@ -80,21 +86,24 @@
...
     "sweetalert2-react-content": "^1.1.0",
+    "swr": "^0.1.5",
     "tslint": "^5.18.0",
     "tslint-config-prettier": "^1.18.0",
     "tslint-plugin-prettier": "^2.0.1",
     "typescript": "^3.5.3"
   },
   "dependencies": {
+    "bson-ext": "^2.0.3",
+    "@types/react-select": "^3.0.0",
```
Comment (lolPants) [inline client/package.json]: This should be a devDependency

File: .gitignore [hunk 3]
```diff
@@ -4,3 +4,5 @@
 # Environment Variables
 *.env
 !example.env
+client/client.iml
```
Comment (lolPants) [inline .gitignore]: These should be clearly labeled with a comment
(see the rest of the file for examples)

File: client/src/ts/routes/Search.tsx [hunk 4]
```diff
@@ -13,42 +15,171 @@ interface IProps {
...
+  {value: 1, label: 'Downloads Desc'},
+  {value: 2, label: 'Rating Asc'},
+  {value: 3, label: 'Rating Desc'},
+  {value: 4, label: 'Plays Asc'},
+  {value: 5, label: 'Plays Desc'},
+];
+
+const Search: FunctionComponent<IProps> = ({pathname, queryStr, replace}) => {
+
+  const reactSelectStyle = {
```
Comment (lolPants) [inline client/src/ts/routes/Search.tsx]: This should have a typedef

File: client/src/ts/routes/Search.tsx [hunk 5]
```diff
@@ -13,42 +15,171 @@ interface IProps {
...
+    singleValue: (provided) => ({
+      ...provided,
+      color: '#eff1f5'
+    }),
+  }
+
+  const [query, setQuery] = useState('*')
+  const [difficulty, setDifficulty] = useState(difficulties.map(d => d.value));
+  const [timeframe, setTimeframe] = useState(1);
+  const [sortBy, setSortBy] = useState(1);
```
Comment (lolPants) [inline client/src/ts/routes/Search.tsx]: All of these state hooks should have generic typings

File: client/src/ts/utils/characteristics.ts [hunk 6]
```diff
@@ -2,11 +2,18 @@ export const parseCharacteristics: (
   chars: IBeatmapCharacteristic[]
 ) => IBeatmapCharacteristic[] = chars => {
   return chars.map(({ name, ...rest }) => {
-    const newName = name
-      .replace(/([A-Z])/g, ' $1')
-      .replace(/^./, str => str.toUpperCase())
-      .trim()
-      .replace(/(  )/g, ' ')
+
+    let newName = ""
```
Comment (lolPants) [inline client/src/ts/utils/characteristics.ts]: Consider refactoring this to not use `let`

File: server/src/index.ts [hunk 7]
```diff
@@ -27,12 +36,27 @@ mongoose
       if (!ELASTIC_DISABLED) {
         axios.get(`http://${ELASTIC_HOST}:${ELASTIC_PORT}`)
         signale.info('Connected to Elasticsearch instance!')
+
+        // if beatmaps index doesn't exist; syncronize
```
Comment (lolPants) [inline server/src/index.ts]: This comment is redundant

File: server/src/index.ts [hunk 8]
```diff
@@ -27,12 +36,27 @@ mongoose
...
+        axios.get(`http://${ELASTIC_HOST}:${ELASTIC_PORT}/beatmaps`).then((response) => {
```
Comment (lolPants) [inline server/src/index.ts]: Consider refactoring to only use `.catch()`

File: server/src/mongo/plugins/mongoosastic.d.ts [hunk 9]
```diff
@@ -78,11 +78,13 @@ declare module 'mongoose' {
 
     public search<S = any>(
       query: any,
+      opts: any,
       callback: (err: Error | undefined, results: ISearchResponse<S>) => any
     ): void
 
     public esSearch<S = any>(
       query: IFullSearch,
+      opts: any,
```
Comment (lolPants) [inline server/src/mongo/plugins/mongoosastic.d.ts]: Please add typings to the best of your ability

File: server/src/mongo/plugins/mongoosastic.d.ts [hunk 10]
```diff
@@ -78,11 +78,13 @@ declare module 'mongoose' {
 
     public search<S = any>(
       query: any,
+      opts: any,
```
Comment (lolPants) [inline server/src/mongo/plugins/mongoosastic.d.ts]: Please add typings to the best of your ability

File: server/src/routes/search.ts [hunk 11]
```diff
@@ -42,14 +42,16 @@ const buildQuery = (query: string, fields: IQueryField[]) =>
     })
     .join(' ')
 
-const elasticSearch = async (query: any, page: number = 0) => {
+const elasticSearch = async (query: any, page: number = 0, sort?: object | null) => {
```
Comment (lolPants) [inline server/src/routes/search.ts]: Please add typings to the best of your ability

File: server/src/routes/search.ts [hunk 12]
```diff
@@ -139,19 +150,126 @@ router.get(
...
+          }
+          break
+        case 5:
+          sort = {
+            "stats.plays": "desc"
+          }
+          break
+      }
+
+    }
```
Comment (lolPants) [inline server/src/routes/search.ts]: Consider breaking down the above into multiple functions.
Avoid `let`.

**Comment by MrCandle** [comment]
is this PR still alive?

**Comment by lolPants** [comment]
closing due to inactivity and impending client rewrite

**Pull request closed without merging**